### ISSUE TITLE
browser.markdown fails at ordered list?

### ISSUE THREAD

**Issue by rayluo** [brython-dev/brython#1736]
Bonjour, Pierre. I notice that the `browser.markdown` does not (yet?) support ordered list.

```python
from browser import markdown
payload = """
1. hello
2. world
"""
html, scripts = markdown.mark(payload)
```

It would encounter an exception.

```
 Traceback (most recent call last):
  File http://192.168.1.8:8000/main.py line 49, in <module>
    html, scripts = markdown.mark(payload)
  File VFS.browser.markdown.py line 154, in mark
    lines[i]='<li>'+lines[i][len(mo.groups()[0]):]
IndexError: tuple index out of range
```

By the way, can you also give a sample of markdown snippet here (and later also add it to [this document page](https://brython.info/static_doc/en/markdown.html)), to demonstrate what input would generate that `scripts`?

---

<sub>Off-topic:
I also attempted to use `brip` to install [this generic markdown module](https://pypi.org/project/Markdown/). The installation was successful, but the `import markdown` failed with this trace:

```
 Traceback (most recent call last):
  File http://192.168.1.8:8000/main.py line 50, in <module>
    import markdown
  File VFS.markdown/__init__.py line 29, in <module>
    from .core import Markdown,markdown,markdownFromFile
  File VFS.markdown.core.py line 26, in <module>
    from . import util
ImportError: cannot import name 'util'
```

Not sure whether Brython's VFS is expected to support `from . import module`. Those are typical errors I encountered when I worked on `brip`. Perhaps you can also help take a look, too.
</sub>

**Comment by PierreQuentel** [comment]
Hello Ray,

Thanks for the report. There were / are a few problems here:
- the Brython module __`browser.markdown`__ used a version of the regular expression engine that failed in the case you have reported. Replacing it by the __`re`__ module in the stdlib fixed the issue (first commit above)
- the generic markdown package does not work because it uses the __`xml`__ module in the standard distribution, which is not supported by Brython. The second commit above changes the reporting of import errors, instead of the one you reported, you should now have a message indicating the reason why `from . import util` fails (this syntax is supported in VFS)
- I edited the source code of this markdown module by removing the reference to __`xml`__ (it is only imported to be put in a list of deprecated features...) and found other issues : module __`importlib.metadata`__ is missing (I have added it), then I came across issue #1740 inside this module

I have noted your suggestion to document what is the `scripts` value in the result of `markdown.mark(src)`.

**Comment by rayluo** [comment]
> Hello Ray,
> 
> Thanks for the report. There were / are a few problems here:

Thanks for replying to each and every sub-questions! I was not supposed to overload an issue with multiple topics. :-)
> 
> * the Brython module **`browser.markdown`** used a version of the regular expression engine that failed in the case you have reported. Replacing it by the **`re`** module in the stdlib fixed the issue (first commit above)

I tested with the latest Brython `master` branch. The fix technically works. However, I remember something, and then a test with some moderate-length markdown content confirmed my fear: the `re` module in stdlib is too slow to be useful. Yeah, that is the ["something" I remember we once discussed with](https://github.com/brython-dev/brython/issues/1592#issuecomment-769040144): "... unbearably slow. I suspect that the re module might be the culprit: the version used in Brython is a pure-Python module, much slower than the C implementation in CPython...". Now we have a dilemma on which `re` to use.

PS: Thanks VERY MUCH for your second commit to generate the correct error! I wish I would have asked you on that earlier, it could make my time on `brip` much more enjoyable. :-)  I won't go into details on `brip` topics here. In the future I might open issues in `brip`'s repo, and discuss with you there.

> I have noted your suggestion to document what is the `scripts` value in the result of `markdown.mark(src)`.

Thanks. Look forward to it.

**Comment by PierreQuentel** [comment]
In fact the re module is no more the pure-Python version used in previous versions of Brython; since release 3.9.2 it is a Javascript module that implements the Python regular expression engine. It is certainly not as fast as CPython's which is in C, but much faster than the pure-Python version.

Can you give an example of a markdown page that takes too long to process ?

**Comment by rayluo** [comment]
> Can you give an example of a markdown page that takes too long to process ?

You can grab the markdown source from [this page](https://github.com/rayluo/first-programming-book/wiki/Preface).

PS: I am attempting to convert my old Python tutorial into an interactive version powered by Brython. Currently, only the first "Home" page is rendered not-too-slow, other pages are unbearable.

**Comment by PierreQuentel** [comment]
You are right, markdown is unbearably slow with the files in your tutorial...

The commit above makes it a little less so, but I think I have found the main culprit : the management of strings.

There is a difference in the way Python and Javascript handle the Unicode characters in the "astral plane" such as 😀. Python counts it as a 1-character string and Javascript as 2-character.

In previous versions I used a specific class for strings with such characters, and plain JS strings for those without. When I rewrote the re engine I chose to use plain strings in all cases, but this means that for instance `__getitem__` requires to scan the whole string each time to see if there are "weird" characters in it; this is what makes markdown slow, and certainly all the code that uses `str.__getitem__`.
